File: src/Java/1. First Course/2. Hole/src/main/refactoringgolf/stack/Stack.java
package refactoringgolf.stack;

import java.util.ArrayList;
import java.util.List;

public class Stack {
	
    private List<Object> elements = new ArrayList<Object>();

    public boolean isEmpty()
    {
        return size() == 0;
    }

    public int size()
    {
        return elements.size();
    }

    public void push(Object element)
    {
        elements.add(element);
    }

    public Object pop()
    {
        if (isEmpty())
            throw new IllegalStateException();
        Object element = elements.get(size() - 1);
        elements.remove(size()-1);
        return element;
    }

    public Object peek()
    {
        return elements.get(size() - 1);
    }

    public boolean contains(Object elementToFind)
    {
        int indexOf = search(elementToFind);
        if (indexOf != -1)
        {
            return true;
        }
        return false;
    }

    public int search(Object elementToFind)
    {
        for (int i = 1; i <= size(); i++)
        {
            if (elementToFind == elements.get(size()-i))
            {
                return i;
            }
        }
        return -1;
    }

    public void replaceAll(Object elementToFind, Object newElement)
    {
        for (int i = size() - 1; i >= 0; i--)
        {
            if (elementToFind == elements.get(i))
            {
                elements.set(i,newElement);
            }
        }
    }
}

File: src/Net/2. Second Course/2. Hole/RefactoringGolf.Stack/Stack.cs
namespace RefactoringGolf.Stack
{
    using System;
    using System.Collections.Generic;

    public class Stack
    {
        private List<object> elements = new List<object>();

        public bool IsEmpty
        {
            get { return Size == 0; }
        }

        public int Size
        {
            get { return elements.Count; }
        }

        public void Push(object element)
        {
            elements.Add(element);
        }

        public object Pop()
        {
            if (IsEmpty)
                throw new InvalidOperationException();
            object element = elements[Size - 1];
            elements.RemoveAt(Size-1);
            return element;
        }

        public object Peek()
        {
            return elements[Size - 1];
        }

        public bool Contains(object elementToFind)
        {
            int indexOf = Search(elementToFind);
            if (indexOf != -1)
            {
                return true;
            }
            return false;
        }

        public int Search(object elementToFind)
        {
            for (int i = 1; i <= Size; i++)
            {
                if (elementToFind == elements[Size-i])
                {
                    return i;
                }
            }
            return -1;
        }

        public void ReplaceAll(object elementToFind, object newElement)
        {
            for (int i = Size - 1; i >= 0; i--)
            {
                if (elementToFind == elements[i])
                {
                    elements[i] = newElement;
                }
            }
        }
    }
}